### PR TITLE
Optimize graph call

### PR DESCRIPTION
**Pull request by imranq2**
Instead of calling mongo for each entity and its descendants, we batch up the calls and make one call for each generation

### CODE REVIEW

File: src/searchParameters/searchParameterHelpers.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+                    return propertyObj.field;
+                }
+            }
+        }
+    }
+    return null;
+}
+
+module.exports = {
+    getFieldNameForSearchParameter: getFieldNameForSearchParameter
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/searchParameters/searchParameterHelpers.js:23]: yep. in GraphHelpers

**Review by DarylCheung — APPROVED** [review]

File: src/operations/graph/oldgraph.js [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+ * @param {string} relatedResourceCollectionName
+ * @param {string} base_version
+ * @param {Resource} parent parent entity
+ * @param {string} host
+ * @param {string | null} filterProperty (Optional) filter the sublist by this property
+ * @param {*} filterValue (Optional) match filterProperty to this value
+ * @param {string} reverse_filter Do a reverse link from child to parent using this property
+ * @return {Promise<[{resource: Resource, fullUrl: string}]>}
+ */
+async function get_reverse_related_resources(db, parentCollectionName, relatedResourceCollectionName, base_version, parent, host, filterProperty, filterValue, reverse_filter) {
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/operations/graph/oldgraph.js:102]: Great idea.  Refactored.

File: src/operations/graph/oldgraph.js [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+                const related_resource_id = relatedResourcePropertyCurrent.reference.replace(collectionName + '/', '');
+
+                /**
+                 * @type {Document | null}
+                 */
+                const found_related_resource = await collection.findOne({id: related_resource_id.toString()});
+                if (found_related_resource) {
+                    // noinspection UnnecessaryLocalVariableJS
+                    entries = entries.concat([{
+                        'fullUrl': `https://${host}/${base_version}/${found_related_resource.resourceType}/${found_related_resource.id}`,
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/operations/graph/oldgraph.js:70]: pass protocol

File: src/operations/graph/oldgraph.js [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+         * @type Resource
+         */
+        for (const relatedResourcePropertyCurrent of relatedResourcePropertyDocuments) {
+            if (filterProperty !== null) {
+                if (relatedResourcePropertyCurrent[`${filterProperty}`] !== filterValue) {
+                    continue;
+                }
+            }
+            entries = entries.concat([{
+                'fullUrl': `https://${host}/${base_version}/${relatedResourcePropertyCurrent.resourceType}/${relatedResourcePropertyCurrent.id}`,
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/operations/graph/oldgraph.js:143]: this is the old code

File: src/operations/graph/oldgraph.js [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+             * @type {[{resource: Resource, fullUrl: string}]}
+             */
+            let entries_for_current_link = [];
+            let link_targets = link.target;
+            for (const target of link_targets) {
+                /**
+                 * @type {string}
+                 */
+                const resourceType = target.type;
+                if (link.path) {
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/operations/graph/oldgraph.js:186]: this is old code that is getting replaced

**Pull request merged**